### PR TITLE
Various fixes after playing around with some scenarios

### PR DESCRIPTION
**Pull request by kgodlewski**

### CODE REVIEW

File: src/neptune_scale/__init__.py [hunk 1]
```diff
@@ -312,6 +346,13 @@ def log(
             ```
 
         """
+
+        # Don't log anything after we've been stopped. This allows continuing the training script
+        # after a non-recoverable error happened.
+        with self._lock:
+            if self._is_closing:
```
[the remark this answers is not in the document]
Comment (Raalsky) [reply src/neptune_scale/__init__.py:353]: + Could we move it after params verification?

File: src/neptune_scale/__init__.py [hunk 2]
```diff
@@ -366,20 +408,27 @@ def _wait(
...
         wait_time = min(sleep_time, timeout) if timeout is not None else sleep_time
-        last_queued_sequence_id = self._operations_queue.last_sequence_id
         last_print_timestamp: Optional[float] = None
 
         while True:
             try:
                 with self._lock:
                     if not self._sync_process.is_alive():
-                        if verbose:
+                        if verbose and not self._is_closing:
```
Comment (Raalsky) [inline src/neptune_scale/__init__.py:417]: We can probably move the flag check `_is_closing` way higher (even in the `while` condition)

File: src/neptune_scale/__init__.py [hunk 3]
```diff
@@ -237,26 +242,55 @@ def resources(self) -> tuple[Resource, ...]:
             self._errors_monitor,
         )
 
-    def _close(self) -> None:
+    def _close(self, *, wait: bool = True) -> None:
         with self._lock:
-            if self._sync_process.is_alive():
+            if self._is_closing:
```
[the remark this answers is not in the document]
Comment (kgodlewski) [reply src/neptune_scale/__init__.py:247]: We need to be sure that we don't start closing from different threads, hence the check is in the crit section

File: src/neptune_scale/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kgodlewski) [reply src/neptune_scale/__init__.py:353]: I'll have another look, I think the critical section here was justified

File: src/neptune_scale/core/components/daemon.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 import threading
 from enum import Enum
 
+from ..logger import logger
```
[the remark this answers is not in the document]
Comment (kgodlewski) [reply src/neptune_scale/core/components/daemon.py:7]: Is it possible to have the static checks enforce that?

File: src/neptune_scale/core/components/operations_queue.py [hunk 5]
```diff
@@ -54,6 +54,7 @@ def enqueue(self, *, operation: RunOperation) -> None:
                 raise ValueError(f"Operation size exceeds the maximum allowed size ({MAX_QUEUE_ELEMENT_SIZE})")
 
             with self._lock:
+                # TODO: should we not block here, and just call the error callback if we were to block?
```
[the remark this answers is not in the document]
Comment (kgodlewski) [reply src/neptune_scale/core/components/operations_queue.py:57]: Yes, I had similar thoughts. Something to be resolved later.

File: src/neptune_scale/__init__.py (hunk 2 above)
Comment (kgodlewski) [reply src/neptune_scale/__init__.py:416]: The goal here was not to issue a warning when it's expected for the sync process to be killed, ie. during closing.

**Review by Raalsky — APPROVED** [review]

**Pull request merged**